File: web/src/utils/formTheme.js
// Tema fill page per-form: {"accent": "#0053db"}.
// Satu warna aksen → varian dark/soft diturunkan otomatis.

export const DEFAULT_ACCENT = '#0053db';

export const THEME_PRESETS = [
  { name: 'Biru Default', accent: '#0053db' },
  { name: 'Emerald', accent: '#059669' },
  { name: 'Violet', accent: '#7c3aed' },
  { name: 'Amber', accent: '#d97706' },
  { name: 'Rose', accent: '#e11d48' },
  { name: 'Cyan', accent: '#0891b2' },
];

const HEX_RE = /^#[0-9a-fA-F]{6}$/;

export function isValidHex(v) {
  return typeof v === 'string' && HEX_RE.test(v);
}

// Kembalikan { accent } yang sudah divalidasi, atau null (= pakai default).
export function normalizeTheme(theme) {
  const accent = theme?.accent;
  if (!isValidHex(accent)) return null;
  return { accent: accent.toLowerCase() };
}

function hexToRgb(hex) {
  const n = parseInt(hex.slice(1), 16);
  return [(n >> 16) & 255, (n >> 8) & 255, n & 255];
}

function mix(hex, target, weight) {
  const [r, g, b] = hexToRgb(hex);
  const m = (c, t) => Math.round(c + (t - c) * weight);
  return `#${((1 << 24) + (m(r, target[0]) << 16) + (m(g, target[1]) << 8) + m(b, target[2])).toString(16).slice(1)}`;
}

// Object style React → CSS vars --accent* (kosong bila tema default).
// Nama var mengikuti konvensi yang sudah dipakai form-fill.css.
export function themeStyle(theme) {
  const t = normalizeTheme(theme);
  if (!t) return {};
  return {
    '--accent': t.accent,
    '--accent-dark': mix(t.accent, [0, 0, 0], 0.18),
    '--accent-soft': mix(t.accent, [255, 255, 255], 0.88),
    '--accent-border': mix(t.accent, [255, 255, 255], 0.72),
    '--accent-ring': mix(t.accent, [255, 255, 255], 0.8),
  };
}

// Aksen efektif (untuk alt text / logic yang butuh nilai konkret).
export function themeAccent(theme) {
  return normalizeTheme(theme)?.accent || DEFAULT_ACCENT;
}

File: mobile/lib/utils/form_theme.dart
// Tema aksen fill page per-form: {"accent": "#0053db"}.
// Cermin web utils/formTheme.js: preset sama, validasi sama, null = default.
// Satu-satunya sumber kebenaran daftar preset + normalisasi di mobile.

/// Aksen default bila theme null (sama seperti web DEFAULT_ACCENT).
const String kFormDefaultAccent = '#0053db';

/// Preset web THEME_PRESETS (nama + hex, urutan sama).
const List<({String name, String accent})> kFormThemePresets = [
  (name: 'Biru Default', accent: '#0053db'),
  (name: 'Emerald', accent: '#059669'),
  (name: 'Violet', accent: '#7c3aed'),
  (name: 'Amber', accent: '#d97706'),
  (name: 'Rose', accent: '#e11d48'),
  (name: 'Cyan', accent: '#0891b2'),
];

final RegExp _hexRe = RegExp(r'^#[0-9a-fA-F]{6}$');

bool isValidFormAccent(String? v) =>
    v is String && _hexRe.hasMatch(v);

/// Kembalikan accent tervalidasi (lowercase) atau null (= default).
/// Menerima String hex langsung atau Map {"accent": ...} dari backend.
String? normalizeFormAccent(dynamic theme) {
  String? raw;
  if (theme is String) {
    raw = theme;
  } else if (theme is Map) {
    raw = theme['accent']?.toString();
  } else {
    return null;
  }
  if (!isValidFormAccent(raw)) return null;
  return raw!.toLowerCase();
}

/// Aksen efektif untuk logika/UI yang butuh nilai konkret.
String formAccentOrDefault(dynamic theme) =>
    normalizeFormAccent(theme) ?? kFormDefaultAccent;
